### PR TITLE
[WebAuthn] Implement conditional creation

### PR DESCRIPTION
**Pull request by g-davidson**
#### ef55edc69e9b28394d7e73facedb868a6ca72da2
<pre>
[WebAuthn] Implement conditional creation
<a href="https://bugs.webkit.org/show_bug.cgi?id=265487">https://bugs.webkit.org/show_bug.cgi?id=265487</a>
<a href="https://rdar.apple.com/113573376">rdar://113573376</a>

Reviewed by Pascoe.

Implement the plumbing to pass the existing CredentialManagement
CredentialMediationRequirement through to the platform. This mostly involved pulling the
mediation requirement up a level from CredentialRequestOptions to being at the top level
of CredentialManagement, which matches the spec and makes up the bulk of the diff.

Pascoe did most of the work here, and I just picked up the last bit.

* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/credentialmanagement/CredentialCreationOptions.h:
* Source/WebCore/Modules/credentialmanagement/CredentialCreationOptions.idl:
* Source/WebCore/Modules/credentialmanagement/CredentialMediationRequirement.idl: Copied from Source/WebCore/Modules/credentialmanagement/CredentialCreationOptions.idl.
* Source/WebCore/Modules/credentialmanagement/CredentialRequestOptions.h:
* Source/WebCore/Modules/credentialmanagement/CredentialRequestOptions.idl:
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.cpp:
(WebCore::CredentialsContainer::get):
(WebCore::CredentialsContainer::isCreate):
* Source/WebCore/Modules/credentialmanagement/MediationRequirement.h: Copied from Source/WebCore/Modules/credentialmanagement/CredentialRequestOptions.h.
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::create):
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.h:
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinatorClient.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::requestWebAuthenticationConditonalMediationRegistration):
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::setDelegate):
(WebKit::UIDelegate::UIClient::requestWebAuthenticationConditonalMediationRegistration):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::AuthenticatorManager::respondReceived):
(WebKit::AuthenticatorManager::initTimeOutTimer):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::configureRegistrationRequestContext):
(WebKit::WebAuthenticatorCoordinatorProxy::contextForRequest):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.cpp:
(WebKit::WebAuthenticatorCoordinatorProxy::makeCredential):
(WebKit::WebAuthenticatorCoordinatorProxy::handleRequest):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.messages.in:
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.cpp:
(WebKit::WebAuthenticatorCoordinator::makeCredential):
(WebKit::WebAuthenticatorCoordinator::getAssertion):
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKWebAuthenticationPanel.mm:
(-[TestWebAuthenticationPanelUIDelegate _webView:requestWebAuthenticationConditionalMediationRegistrationForUser:completionHandler:]):

Canonical link: <a href="https://commits.webkit.org/271353@main">https://commits.webkit.org/271353@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/83802f2c24789fa4bb685d1e939092a0bd53664e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27974 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6612 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29276 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30504 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25516 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28469 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8607 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4003 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/25281 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28241 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5368 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24034 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4623 "Found 1 new API test failure: /WPE/TestWebKitWebView:/webkit/WebKitWebView/notification (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4805 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25041 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31193 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25580 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25477 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31067 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4809 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2975 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28877 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/notification (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6344 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6739 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5254 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5295 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/Modules/credentialmanagement/MediationRequirement.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace WebCore {
+
+enum class MediationRequirement : uint8_t { Silent, Optional, Required, Conditional };
+
+} // namespace WebCore
+
+namespace WTF {
+
+template<> struct EnumTraits<WebCore::MediationRequirement> {
```
Comment (rr-codes) [inline Source/WebCore/Modules/credentialmanagement/MediationRequirement.h]: I _think_ EnumTraits is no longer needed in favor of the new generated enum serialization mechanism, but I could be wrong

File: Source/WebKit/UIProcess/API/APIUIClient.h [hunk 2]
```diff
@@ -217,6 +217,11 @@ class UIClient {
     {
         completionHandler(true);
     }
+
+    virtual void requestWebAuthenticationConditonalMediationRegistration(const WTF::String&&, CompletionHandler<void(bool)>&& completionHandler)
```
Comment (rr-codes) [inline Source/WebKit/UIProcess/API/APIUIClient.h]: is the `WTF::` needed here? it typically isn't

Comment (rr-codes) [reply Source/WebKit/UIProcess/API/APIUIClient.h]: also is a `const` `&&` a thing?

File: Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.cpp [hunk 3]
```diff
@@ -74,6 +74,12 @@ void WebAuthenticatorCoordinatorProxy::handleRequest(WebAuthenticationRequestDat
 {
     auto origin = API::SecurityOrigin::create(data.frameInfo.securityOrigin.protocol(), data.frameInfo.securityOrigin.host(), data.frameInfo.securityOrigin.port());
 
+    bool shouldRequestConditionalRegistration = std::holds_alternative<PublicKeyCredentialCreationOptions>(data.options) && data.mediation == MediationRequirement::Conditional;
```
Comment (pascoej) [inline Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.cpp:77]:
```suggestion
String username;
if (auto* createOptions = dynamicDowncast<PublicKeyCredentialCreationOptions>(data.options); options && data.mediation == MediationRequirement::Conditional)
    username = createOptions->name;
```

Comment (pascoej) [reply Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.cpp:77]: I think this would also work, though I don't think it's any more clear so this isn't even a nit. (please ignore)

File: Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.cpp [hunk 4]
```diff
@@ -66,17 +67,17 @@ WebAuthenticatorCoordinator::WebAuthenticatorCoordinator(WebPage& webPage)
...
     auto webFrame = WebFrame::fromCoreFrame(frame);
     if (!webFrame)
         return;
 
     auto isProcessingUserGesture = processingUserGesture(frame, webFrame->frameID());
-    m_webPage.sendWithAsyncReply(Messages::WebAuthenticatorCoordinatorProxy::MakeCredential(webFrame->frameID(), webFrame->info(), hash, options, isProcessingUserGesture), WTFMove(handler));
+    m_webPage.sendWithAsyncReply(Messages::WebAuthenticatorCoordinatorProxy::MakeCredential(webFrame->frameID(), webFrame->info(), hash, options, mediation, isProcessingUserGesture), WTFMove(handler));
 }
 
-void WebAuthenticatorCoordinator::getAssertion(const LocalFrame& frame, const SecurityOrigin&, const Vector<uint8_t>& hash, const PublicKeyCredentialRequestOptions& options, MediationRequirement mediation, const ScopeAndCrossOriginParent& scopeAndCrossOriginParent, RequestCompletionHandler&& handler)
```
Comment (pascoej) [inline Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.cpp:79]: Though I'm think I've written it in other places in this file, we are `using namespace WebCore;` at the top so we don't need the prefix. Not the case in the header though.

File: Source/WebCore/Modules/credentialmanagement/MediationRequirement.h (hunk 1 above)
Comment (g-davidson) [reply Source/WebCore/Modules/credentialmanagement/MediationRequirement.h]: As someone who contributes to WebKit ~once/year, can you tell me more about this? :)

File: Source/WebKit/UIProcess/API/APIUIClient.h (hunk 2 above)
Comment (g-davidson) [reply Source/WebKit/UIProcess/API/APIUIClient.h]: Yes the `WTF::` is needed and matches the rest of the file. Good catch with the `const` though.

**Comment by pascoej** [comment]
You're going to have to add the new source files to some sources.lists files to fix the wincairo and gtk builds. I recommend full text searching for a nearby similar file, and adding the new file path to all those places (note: the sources files for other platforms aren't indexed by Xcode by default.)

File: Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.cpp (hunk 3 above)
Comment (g-davidson) [reply Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.cpp:77]: That's neat! I still think I need the bool because I check it again lower down.

File: Source/WebCore/Modules/credentialmanagement/MediationRequirement.h (hunk 1 above)
Comment (pascoej) [reply Source/WebCore/Modules/credentialmanagement/MediationRequirement.h]: A new generator for serialization handles what the `EnumTraits` line here did. You can see an example of the change needed for `AttestationConveyancePreference` in https://github.com/WebKit/WebKit/pull/7694/files#diff-b7c9fc545913900a0089417beb52d397de465e8e1c044c1718efd2eb05307ab1. 

tl;dr: remove this line / everything in WTF namespace here. Then add 
`#if ENABLE(WEB_AUTHN)
enum class MediationRequirement : uint8_t { Silent, Optional, Required, Conditional };
#endif
`

in `WebCoreArgumentCoders.serialization.in`

Comment (g-davidson) [reply Source/WebCore/Modules/credentialmanagement/MediationRequirement.h]: Neat, thank you!

**Review by pascoej — APPROVED** [review]

**Comment by pascoej** [comment]
Thank you Garrett!
